kernel/byterun/coq_values.c

```diff
@@ -96,3 +96,23 @@ value coq_tcode_array(value tcodes) {
   }
   CAMLreturn(res);
 }
+
+/* Low-level representation of accumulators for native compilation */
+
+static code_t coq_accumulator_code;
+static int coq_accumulator_init = 0;
+
+value coq_native_register_accu(value v) {
+  if (coq_accumulator_init == 0) {
+    coq_accumulator_init = 1;
+    coq_accumulator_code = Code_val(v);
+  }
+  return Val_unit;
+}
+
+value coq_native_is_accu(value v) {
+  if (Tag_val(v) == Closure_tag) {
+    if (Code_val(v) == coq_accumulator_code) return Val_true;
```

Comment:
FYI C has short circuited "logical" conjunction

Reply:
Yes, I know :) This code is not correct anyways, this PR is a draft for a good reason.

Reply:
Anyway, the inner test is useless. If it were to fail, the value will get matched upon (while it has a closure tag) and Coq will segfault.

That is the equivalent of the following macro in the bytecode interpreter:
```c
#define Is_accu(v) (Is_block(v) && Tag_val(v) == Closure_tag)
```

By the way, I am a bit surprised you are not testing `Is_block(v)` in your version. Are you really sure the value is a block there?

Reply:
> Anyway, the inner test is useless.

Not for all calls to `is_accu`. This is only true for inductive accumulators encountered in match nodes. In the reification code `kind_of_value`, we don't have this invariant since we don't have the type around, and we can produce functions `Vfun` with a closure tag.

Reply:
Then I suggest making two functions, one for execution and one for reification. (Especially since I am still under the impression that the execution one is supposed to test for `Is_block`.)

Reply:
(Note that there is an ulterior motive for this suggestion. If we can make it work by just testing for closure rather than full-blown accumulator, it will be easier to convince OCaml developers to optimize for this case, e.g., builtin `Obj.is_closure`.)

Reply:
You are right that `Is_block` is needed though, I naively thought that `Tag_val` was checking for integers.

Reply:
> it will be easier to convince OCaml developers to optimize for this case

I'd rather compile to malfunction and have a low-level enough language of patterns to jump on the tag directly.

Reply:
That is my point. The following seems just fine as a low-level language of patterns:
```ocaml
match foo with
| x when Obj.is_closure x -> ...
```

Reply:
Hm, then we would also need language support for ADT containing functions. If OCaml is clever enough it should rule out this branch statically since no constructor can ever be a function.

Reply:
(Native compilation is likely to be a huge UB anyways, I really think targetting OCaml and playing fast and loose with the type system is a footgun waiting to be shot in our face when we start using flambda.)

Reply:
> If OCaml is clever enough it should rule out this branch statically since no constructor can ever be a function.

Note that OCaml currently allows to perform a match on a polymorphic values, so it cannot be too clever. Just because a value is passed to `match` does not mean that it is a constructor.
```ocaml
let f (g:'a -> bool) x =
  match x with
  | x when g x -> true
  | x -> false
```

Now, what if the match also contains proper branches? I guess that is one of those things that were never officially stated (and for which flambda might indeed be playing fast and loose).